### PR TITLE
Changing the data frequency for Updates and other changes.

### PR DESCRIPTION
**Pull request by vrdmr**
Changed the following:
1. Changing the run_interval to 3h, and removed the forcefully sending.
2. Adding patch_management_immediate_run.conf file which is a copy
of the patch_management.conf without run_interval to start as soon as
agent starts. Different tag used for the same.
3. If _BuildDate_ and _InstalledOn_ is either nil or **BuildDate**, **InstalelledOn**, 
then send nil.
4. Also made changes in the tests.

@Microsoft/omsagent-devs @stasku Could you please review the changes?

### CODE REVIEW

File: installer/conf/omsagent.d/patch_management_immediate_run.conf [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<source>
+  type exec
+  tag oms.patch_management_immediate_run
+  command /opt/microsoft/omsconfig/Scripts/PerformInventory.py --InMOF /etc/opt/microsoft/omsagent/conf/omsagent.d/patch_management_inventory.mof --OutXML /etc/opt/omi/conf/omsconfig/configuration/CompletePackageInventory.xml > /dev/null && cat /etc/opt/omi/conf/omsconfig/configuration/CompletePackageInventory.xml
```
Comment (robbiezhang) [inline installer/conf/omsagent.d/patch_management_immediate_run.conf]: is it by purpose to let it run when agent restarts? do you need to distinguish the 2 types of data in the service side?

will it cause race condition with the scheduled one, as they use the same output file?

File: source/code/plugins/patch_management_lib.rb [hunk 2]
```diff
@@ -150,8 +150,11 @@ def self.availableUpdatesXMLtoHash(availableUpdatesXML, os_short_name)
         ret["CollectionName"] = availableUpdatesHash["Name"] + @@delimiter + 
                                 availableUpdatesHash["Version"] + @@delimiter + os_short_name
         ret["PackageName"] = availableUpdatesHash["Name"]
+        ret["Architecture"] = availableUpdatesHash.key?("Architecture") ? availableUpdatesHash["Architecture"] : nil
         ret["PackageVersion"] = availableUpdatesHash["Version"]
-        ret["Timestamp"] = OMS::Common.format_time(availableUpdatesHash["BuildDate"].to_i)
+        ret["Timestamp"] = (availableUpdatesHash["BuildDate"] == nil or
```
Comment (robbiezhang) [inline source/code/plugins/patch_management_lib.rb]: use .nil? to check nil

File: source/code/plugins/patch_management_lib.rb [hunk 3]
```diff
@@ -164,8 +167,11 @@ def self.installedPackageXMLtoHash(packageXML, os_short_name)
         ret["CollectionName"] = packageHash["Name"] + @@delimiter + 
                                 packageHash["Version"] + @@delimiter + os_short_name
         ret["PackageName"] = packageHash["Name"]
+        ret["Architecture"] = packageHash.key?("Architecture") ? packageHash["Architecture"] : nil
         ret["PackageVersion"] = packageHash["Version"]
-        ret["Timestamp"] = OMS::Common.format_time(packageHash["InstalledOn"].to_i)
+        ret["Timestamp"] = (packageHash["InstalledOn"] == nil or
```
Comment (robbiezhang) [inline source/code/plugins/patch_management_lib.rb]: use .nil?

**Comment by robbiezhang** [comment]
Add test for BuildDate and InstalledOn

File: installer/conf/omsagent.d/patch_management_immediate_run.conf (hunk 1 above)
Comment (vrdmr) [inline installer/conf/omsagent.d/patch_management_immediate_run.conf]: Yes, as out run_interval is now set to 3 hours (earlier, it was 12 hours), the users do not see any data as soon as they onboard. From fluentd forums, the preferred way was to crontab the python separately and use in_tail (https://groups.google.com/forum/#!topic/fluentd/Yq3BRYxYQuI), but then we lose the flexibility to restart if anything happens to the python script as well as the installers then woudl be changed, which was not feasible.

The data is the same from oms.patch_management_immediate_run and oms.patch_management. Also, the oms.patch_management_immediate_run should run only once based on the fluentd code.

File: source/code/plugins/patch_management_lib.rb (hunk 2 above)
Comment (vrdmr) [inline source/code/plugins/patch_management_lib.rb]: Yes. I'll do that.

File: source/code/plugins/patch_management_lib.rb (hunk 3 above)
Comment (vrdmr) [inline source/code/plugins/patch_management_lib.rb]: Yes.

**Comment by vrdmr** [comment]
Added extra test-cases for nil and correct BuildDate/InstalledOn. The unittests are now added and works 100% successful.

File: source/code/plugins/patch_management_lib.rb [hunk 4]
```diff
@@ -164,10 +168,13 @@ def self.installedPackageXMLtoHash(packageXML, os_short_name)
         ret["CollectionName"] = packageHash["Name"] + @@delimiter + 
                                 packageHash["Version"] + @@delimiter + os_short_name
         ret["PackageName"] = packageHash["Name"]
-        ret["PackageVersion"] = packageHash["Version"]
-        ret["Timestamp"] = OMS::Common.format_time(packageHash["InstalledOn"].to_i)
+        ret["Architecture"] = packageHash.key?("Architecture") ? packageHash["Architecture"] : nil
+        ret["PackageVersion"] = packageHash["Version"]  
+        ret["Timestamp"] = (packageHash["InstalledOn"].strip.nil? or
```
Comment (robbiezhang) [inline source/code/plugins/patch_management_lib.rb]: you should not call .strip here, since it's a nil check

File: source/code/plugins/patch_management_lib.rb [hunk 5]
```diff
@@ -150,9 +150,13 @@ def self.availableUpdatesXMLtoHash(availableUpdatesXML, os_short_name)
         ret["CollectionName"] = availableUpdatesHash["Name"] + @@delimiter + 
                                 availableUpdatesHash["Version"] + @@delimiter + os_short_name
         ret["PackageName"] = availableUpdatesHash["Name"]
+        ret["Architecture"] = availableUpdatesHash.key?("Architecture") ? availableUpdatesHash["Architecture"] : nil
         ret["PackageVersion"] = availableUpdatesHash["Version"]
-        ret["Timestamp"] = OMS::Common.format_time(availableUpdatesHash["BuildDate"].to_i)
-        ret["Repository"] = availableUpdatesHash.key?("Repository") ? availableUpdatesHash["Repository"] : ""
+        ret["Timestamp"] = (availableUpdatesHash["BuildDate"].strip == nil or
```
Comment (robbiezhang) [inline source/code/plugins/patch_management_lib.rb]: availableUpdatesHash["BuildDate"].nil?

**Comment by vrdmr** [comment]
Changes done.

**Comment by robbiezhang** [comment]
Signoff

**Comment by vrdmr** [comment]
Merged the two patch management related confs into one. Tested. Now merging the changes.
